src/pydvl/influence/frameworks/torch_differentiable.py

```diff
@@ -26,9 +27,164 @@ def flatten_gradient(grad):
+        bi_cg, _ = solve_cg(reg_hvp, bi, **inversion_method_kwargs)
+        batch_cg[idx] = bi_cg
+    return batch_cg
+
+
+def solve_cg(
+    hvp: Callable[[torch.Tensor], torch.Tensor],
+    b: torch.Tensor,
+    x0: Optional[torch.Tensor] = None,
+    rtol: float = 1e-7,
+    atol: float = 1e-7,
+    maxiter: Optional[int] = None,
+):
+    """Conjugate gradient solver for the Hessian vector product
+
+    :param hvp: a Callable Hvp, operating with tensors of size N
+    :param b: a tensor of shape [N]
+    :param x0: initial guess for hvp
+    :param rtol: maximum relative tolerance of result
+    :param atol: absolute tolerance of result
+    :param maxiter: maximum number of iterations. If None, defaults to 10*len(y)
+    """
+    if x0 is None:
+        x0 = torch.clone(b)
+    if maxiter is None:
+        maxiter = len(b) * 10
+
+    y_norm = torch.sum(torch.matmul(b, b)).item()
+    stopping_val = max([rtol**2 * y_norm, atol**2])
+
+    x = x0
+    p = r = (b - hvp(x)).squeeze()
+    gamma = torch.sum(torch.matmul(r, r)).item()
+    optimal = False
+
+    for k in range(maxiter):
+        Ap = hvp(p).squeeze()
+        alpha = gamma / torch.sum(torch.matmul(p, Ap)).item()
+        x += alpha * p
+        r -= alpha * Ap
+        gamma_ = torch.sum(torch.matmul(r, r)).item()
+        beta = gamma_ / gamma
+        gamma = gamma_
+        p = r + beta * p
+
+        if gamma < stopping_val:
+            optimal = True
+            break
+
+    info = {"niter": k, "optimal": optimal}
+    return x, info
+
+
+def solve_lissa(
+    model: TwiceDifferentiable,
+    training_data: DataLoader,
+    b: torch.Tensor,
+    lam: float = 0,
+    progress: bool = True,
+    maxiter: int = 1000,
+    damp: float = 0,
+    scale: float = 10,
+):
+    """
+    It uses LISSA, Linear time Stochastic Second-Order Algorithm, to calculate the
+    inverse of the HVP. More precisely, it finds x s.t. $Hx = b$, with $H$ being
+    the model hessian.
+    This is done by iteratively approximating H through
+    $$
+    H^{-1}_{j+1} b = b + (I - H) \ H^{-1}_j b
+    $$
+    where I is the identity matrix. Additional damping and scaling factors are
+    applied to help convergence. More info can be found in
+    :footcite:t:`koh_understanding_2017`
+
+    :param model: A model wrapped in the TwiceDifferentiable interface.
+    :param training_data: A DataLoader containing the training data.
+    :param b:
+    :param lam: regularization of the hessian
+    :param maxiter: maximum number of iterations,
+    :param damp: damping factor, defaults to 0 for no damping
+    :param scale: scaling factor, defaults to 10
+    :param progress: If True, display progress bars.
+
+    :return: A matrix of shape [NxP] with each line being a solution of $Ax=b$.
+    """
+    h_estimate = torch.clone(b)
+    for _ in maybe_progress(range(maxiter), progress, desc="Lissa"):
+        x, y = next(iter(training_data))
```

Comment:
Calling `iter()` on training_data on each iteration recreates the iterator each time.
I think creating it outside the loop would be better `iterator = iter(training_data)`

Reply:
What we want to do here is to sample a batch with different random points every time. I searched around and there seems to be no clean way to do it apart for re-creating the iterator every time.

Reply:
I see. Makes sense. Thanks for the explanation.